### PR TITLE
Scoop filter

### PR DESCRIPTION
**Pull request by laclark22**
An exported patient will only include data relevant to the measure(s) being exported (scoop and filter)

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/CYPRESS-243
- [x] Internal ticket links to this PR https://jira.mitre.org/browse/CYPRESS-243
- [x] Code diff has been done and been reviewed
- [ ] Tests are included and test edge cases
- [ ] Tests have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

**Comment by rbclark** [comment]
@laclark22 It looks like you need to run bundle install and then commit the Gemfile.lock before travis will pass.

File: lib/cypress/scoop_and_filter.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+module Cypress
+  class ScoopAndFilter
+    def initialize(measures)
+      @relevant_codes = codes_in_measures(measures)
+      @demographic_oids = ['2.16.840.1.113883.10.20.28.3.55', '2.16.840.1.113883.10.20.28.3.59', '2.16.840.1.113883.10.20.28.3.56']
+      @hqmf_oids_for_measures = []
```
Comment (rbclark) [inline lib/cypress/scoop_and_filter.rb]: This line seems like it might be unnecessary since it is set to something different on the next line.

File: lib/cypress/patient_zipper.rb [hunk 2]
```diff
@@ -54,6 +54,7 @@ class PatientZipper
     def self.zip(file, patients, format)
       patients = apply_sort_to patients
       measures, sd, ed = measure_start_end(patients)
+      my_scoop_and_filter = Cypress::ScoopAndFilter.new(measures)
```
Comment (rbclark) [inline lib/cypress/patient_zipper.rb]: I'm slightly confused what the `my_` prefix means in this context.

File: lib/cypress/scoop_and_filter.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      end
+      valuesets.each do |valueset|
+        code_list.concat(valueset.concepts)
+      end
+      code_list.map { |cl| { code: cl.code, codeSystem: cl.code_system_name } }
+    end
+
+    def get_all_hqmf_oids_definition_and_status(measures)
+      all_hqmf_oids = []
+      measures.each do |measure|
```
Comment (rbclark) [inline lib/cypress/scoop_and_filter.rb]: It might be better to replace this whole function with the following:
```
def get_all_hqmf_oids_definition_and_status(measures)
  measures.collect do |measure|
    measure.source_data_criteria.collect do |_key, criteria|
      get_all_hqmf_oids(criteria['definition'], criteria['status'])
    end
  end.flatten.uniq
end
```

I ran a crude speed test and it seems collect is a lot faster overall and it over. Below are the speed test results with the old and new methods:

**Old:**
```
       user     system      total        real
   0.070000   0.000000   0.070000 (  0.072991)

       user     system      total        real
   0.060000   0.000000   0.060000 (  0.061780)

       user     system      total        real
   0.010000   0.000000   0.010000 (  0.014338)
```

**New:**
```
       user     system      total        real
   0.020000   0.000000   0.020000 (  0.032177)

       user     system      total        real
   0.020000   0.000000   0.020000 (  0.015087)

       user     system      total        real
   0.020000   0.000000   0.020000 (  0.017952)
```

File: lib/cypress/scoop_and_filter.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      all_hqmf_oids.uniq!
+    end
+
+    def get_all_hqmf_oids(definition, status)
+      hqmf_oids = []
+      vesion_negation_combinations = [{ version: 'r1', negation: false },
+                                      { version: 'r1', negation: true },
+                                      { version: 'r2', negation: false },
+                                      { version: 'r2cql', negation: false }]
+      vesion_negation_combinations.each do |obj|
```
Comment (rbclark) [inline lib/cypress/scoop_and_filter.rb]: Ditto on collect vs each here as well.

File: lib/cypress/scoop_and_filter.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      @demographic_oids = ['2.16.840.1.113883.10.20.28.3.55', '2.16.840.1.113883.10.20.28.3.59', '2.16.840.1.113883.10.20.28.3.56']
+      @hqmf_oids_for_measures = []
+      @hqmf_oids_for_measures = get_all_hqmf_oids_definition_and_status(measures) - @demographic_oids
+    end
+
+    # return an array of all of the concepts in all of the valueset for the measure
+    def codes_in_measures(measures)
+      valuesets = []
+      code_list = []
+      measures.each do |measure|
```
Comment (rbclark) [inline lib/cypress/scoop_and_filter.rb]: Ditto on collect vs each here as well.

File: lib/cypress/scoop_and_filter.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        measure['value_set_oid_version_objects'].each do |valueset|
+          valuesets.concat(HealthDataStandards::SVS::ValueSet.where(oid: valueset.oid, version: valueset.version).to_a)
+        end
+      end
+      valuesets.each do |valueset|
```
Comment (rbclark) [inline lib/cypress/scoop_and_filter.rb]: Ditto on collect vs each here as well.

File: lib/cypress/scoop_and_filter.rb (hunk 3 above)
Comment (rbclark) [reply lib/cypress/scoop_and_filter.rb]: Also some of these return nil, and that nil seems to get included in the returned information. Is this intended behavior?
